x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/aggs/categorization/InternalCategorizationAggregation.java

```diff
@@ -0,0 +1,436 @@
+                TextCategorization group = categorizationTokenTree.parseLogLineConst(hash.getIds(delayedBucket.key.keyAsTokens()));
+                if (group == null) {
+                    throw new AggregationExecutionException(
+                        "Unexpected null categorization group for bucket [" + delayedBucket.key.asString() + "]"
+                    );
+                }
+                BytesRef[] categoryTokens = hash.getDeeps(group.getCategorization());
+
+                BucketKey key = reduceContext.isFinalReduce() ?
+                    BucketKey.withCollapsedWildcards(categoryTokens) :
+                    new BucketKey(categoryTokens);
+                mergedBuckets.computeIfAbsent(
+                        key,
+                        k -> new DelayedCategorizationBucket(k, new ArrayList<>(delayedBucket.toReduce.size()), 0L)
+                    ).add(delayedBucket);
+            }
+
+            final int size = reduceContext.isFinalReduce() == false ? mergedBuckets.size() : Math.min(requiredSize, mergedBuckets.size());
+            final PriorityQueue<Bucket> pq = new BucketCountPriorityQueue(size);
+            for (Map.Entry<BucketKey, DelayedCategorizationBucket> keyAndBuckets : mergedBuckets.entrySet()) {
+                final BucketKey key = keyAndBuckets.getKey();
+                DelayedCategorizationBucket bucket = keyAndBuckets.getValue();
+                Bucket newBucket = bucket.reduce(key, reduceContext);
+                if ((newBucket.docCount >= minDocCount) || reduceContext.isFinalReduce() == false) {
+                    Bucket removed = pq.insertWithOverflow(newBucket);
+                    if (removed == null) {
+                        reduceContext.consumeBucketsAndMaybeBreak(1);
+                    } else {
+                        reduceContext.consumeBucketsAndMaybeBreak(-countInnerBucket(removed));
+                    }
+                } else {
+                    reduceContext.consumeBucketsAndMaybeBreak(-countInnerBucket(newBucket));
+                }
+            }
+            Bucket[] bucketList = new Bucket[pq.size()];
+            for (int i = pq.size() - 1; i >= 0; i--) {
+                bucketList[i] = pq.pop();
+            }
+            // Keep the top categories top, but then sort by the key for those with duplicate counts
+            if (reduceContext.isFinalReduce()) {
+                Arrays.sort(bucketList, Comparator.comparing(Bucket::getDocCount).reversed().thenComparing(Bucket::getRawKey));
+            }
+            return new InternalCategorizationAggregation(
+                name,
+                requiredSize,
+                minDocCount,
+                maxUniqueTokens,
+                maxMatchTokens,
+                similarityThreshold,
+                metadata,
+                Arrays.asList(bucketList)
+            );
+        } catch (IOException ex) {
+            throw new UncheckedIOException(ex);
```

Comment:
I think you could drop this if `CategorizationBytesRefHash` is `Releasable` instead of `Closeable`.